File: sturdr/dsp/tracking.py
"""**tracking.py**

======  ============================================================================================
file    sturdr/dsp/tracking.py
brief   Satellite scalar tracking methods.
date    October 2024
refs    1. "Understanding GPS/GNSS Principles and Applications", 3rd Edition, 2017 
            - Kaplan & Hegarty
        2. "Are PLLs Dead? A Tutorial on Kalman Filter-Based Techniques for Digital Carrier Syncronization" 
            - Vila-Valls, Closas, Navarro, Fernandez-Prades
        3. "A Software-Defined GPS and Galileo Receiver: A Single-Frequency Approach", 2007
            - Borre, Akos, Bertelsen, Rinder, Jensen
        4. "Position, Navigation, and Timing Technologies in the 21st Century", Volume 1, 2021
            - Morton, Diggelen, Spilker Jr., Parkinson
======  ============================================================================================
"""

import numpy as np
from numba import njit, float64
from numba.experimental import jitclass
from sturdr.dsp.discriminator import *

# ===== LOOP FILTERS ============================================================================= #
# See [Kaplan & Hegarty, 2017] pg. 464

@njit(cache=True, fastmath=True)
def NaturalFrequency(bw: np.double, order: int):
    """
    Calculate the natual radian frequency of the loop filter given the noise bandwidth

    Parameters
    ----------
    bw : np.double
        Noise bandwidth [Hz]
    order : int
        Loop filter order (2 or 3)

    Returns
    -------
    w0 : np.double
        Loop filter's natural radian frequency [rad/s]

    Raises
    ------
    ValueError
        Invlaid loop filter order
    """
    a = 1.4142 # 0.7071
    b = 1.1
    c = 2.4
    
    if order == 1:
        return bw / 4
    if order == 2:
        return bw * (4 * a**2) / (a**2 + 1)
    elif order == 3:
        return bw * (4 * (b*c - 1)) / (b*c**2 + (b**2 - c))
    else:
        raise ValueError(f"Loop Filter order of {order} is not valid.")

@njit(cache=True, fastmath=True)
def FLLassistedPLL_2rdOrder(phase_err: np.double, 
                            freq_err: np.double, 
                            vel_accumulator: np.double, 
                            T: np.double, 
                            w0p: np.double, 
                            w0f: np.double):
    """
    2nd order PLL/DLL assisted by 1st order FLL Digital Loop Filter

    Parameters
    ----------
    phase_err : np.double
        Phase error input (discriminator) [rad]
    freq_err : np.double
        Frequency error input (discriminator) [rad/s]
    vel_accumulator : np.double
        Velocity accumulator memory (previous value)
    T : np.double
        Coherent integration time [s]
    w0p : np.double
        Natural radian frequency of the PLL [rad/s]
    w0f : np.double
        Natural radian frequency of the FLL [rad/s]

    Returns
    -------
    output : np.double
        Doppler/Frequency esitmate from the Digital Loop Filter
    vel_accumulator : np.double
        Updated velocity accumulator memory
    """
    a = 1.4142 # 0.7071
    
    # velocity accumulator
    update = vel_accumulator + T * (phase_err * w0p**2 + freq_err * w0f)
    output = (vel_accumulator + update) * 0.5 + (phase_err * a * w0p)
    
    return output, update

@njit(cache=True, fastmath=True)
def FLLassistedPLL_3rdOrder(phase_err: np.double, 
                            freq_err: np.double, 
                            acc_accumulator: np.double, 
                            vel_accumulator: np.double, 
                            T: np.double, 
                            w0p: np.double, 
                            w0f: np.double):
    """
    3rd order PLL/DLL assisted by 2nd order FLL Digital Loop Filter

    Parameters
    ----------
    phase_err : np.double
        Phase error input (discriminator) [rad]
    freq_err : np.double
        Frequency error input (discriminator) [rad/s]
    acc_accumulator : np.double
        Acceleration accumulator memory (previous value)
    vel_accumulator : np.double
        Velocity accumulator memory (previous value)
    T : np.double
        Coherent integration time [s]
    w0p : np.double
        Natural radian frequency of the PLL [rad/s]
    w0f : np.double
        Natural radian frequency of the FLL [rad/s]

    Returns
    -------
    output : np.double
        Doppler/Frequency esitmate from the Digital Loop Filter
    acc_accumulator : np.double
        Updated acceleration accumulator memory
    vel_accumulator : np.double
        Updated velocity accumulator memory
    """
    
    a = 1.4142 # 0.7071
    b = 1.1
    c = 2.4
    
    jitter = acc_accumulator + T * (w0p**3*phase_err + w0f**2*freq_err)
    freq = vel_accumulator + T * (0.5*(acc_accumulator + jitter) + b*w0p**2*phase_err + a*w0f*freq_err)
    output = 0.5*(vel_accumulator + freq) + c*w0p*phase_err
    
    return output, jitter, freq

@njit(cache=True, fastmath=True)
def PLL_2ndOrder(phase_err: np.double, vel_accumulator: np.double, T: np.double, w0d: np.double):
    """
    2nd order Phase/Delay Lock Loop

    Parameters
    ----------
    phase_err : np.double
        Phase error input (discriminator) [rad or chip]
    vel_accumulator : np.double
        Velocity accumulator memory (previous value)
    T : np.double
        Coherent integration time [s]
    w0d : np.double
        Natural radian frequency of the DLL [rad/s]

    Returns
    -------
    _type_
        _description_
    """
    a = 1.4142 # 0.7071
    
    update = vel_accumulator + T * w0d**2*phase_err
    output = 0.5*(vel_accumulator + update) + a*w0d*phase_err 
    
    return output, update

@njit(cache=True, fastmath=True)
def PLL_3rdOrder(phase_err: np.double, 
                 acc_accumulator: np.double, 
                 vel_accumulator: np.double, 
                 T: np.double, 
                 w0p: np.double):
    """
    3nd order PLL/DLL Digital Loop Filter

    Parameters
    ----------
    phase_err : np.double
        Phase error input (discriminator) [rad or chip]
    acc_accumulator : np.double
        Acceleration accumulator memory (previous value)
    vel_accumulator : np.double
        Velocity accumulator memory (previous value)
    T : np.double
        Coherent integration time [s]
    w0p : np.double
        Natural radian frequency of the PLL [rad/s]

    Returns
    -------
    output : np.double
        Doppler/Frequency esitmate from the Digital Loop Filter
    acc_accumulator : np.double
        Updated acceleration accumulator memory
    vel_accumulator : np.double
        Updated velocity accumulator memory
    """
    b = 1.1
    c = 2.4
    
    jitter = acc_accumulator + T*w0p**3*phase_err
    freq = vel_accumulator + T * (0.5*(acc_accumulator + jitter) + b*w0p**2*phase_err)
    output = 0.5*(vel_accumulator + freq) + c*w0p*phase_err
    
    return output, jitter, freq

# ===== KALMAN FILTER ============================================================================ #

# spec = [
#     ('x', float64[:]),
#     ('u', float64[:]),
#     ('P', float64[:]),
#     ('A', float64[:]),
#     ('B', float64[:]),
#     ('C', float64[:]),
#     ('Q', float64[:]),
#     ('R', float64[:]),
#     ('T', float64),
#     ('cn0', float64),
#     ('kappa', float64),
#     ('w0d', float64),
#     ('w0p', float64),
#     ('w0f', float64),
# ]
# @jitclass(spec=spec)
class TrackingKF:
    __slots__ = 'x', 'u', 'P', 'A', 'B', 'C', 'Q', 'R', 'T', 'cn0', 'kappa', 'w0d', 'w0p', 'w0f'
    x     : np.ndarray[np.double]
    u     : np.ndarray[np.double]
    P     : np.ndarray[np.double]
    A     : np.ndarray[np.double]
    B     : np.ndarray[np.double]
    C     : np.ndarray[np.double]
    Q     : np.ndarray[np.double]
    R     : np.ndarray[np.double]
    T     : np.double
    cn0   : np.double
    kappa : np.double
    w0d   : np.double
    w0p   : np.double
    w0f   : np.double
    
    def __init__(self, 
                 w0p: np.double=0.0,
                 w0f: np.double=0.0,
                 w0d: np.double=0.0,
                 doppler: np.double=0.0, 
                 carrier_phase: np.double=0.0, 
                 code_phase: np.double=0.0, 
                 nominal_carrier_freq: np.double=2.0*np.pi*1575.42e6,
                 nominal_code_freq: np.double=1.023e6,
                 intermediate_freq: np.double=0.0,
                 cn0: np.double=5000.0,
                 T: np.double=0.001):
        """
        Constructor for TrackingKF class

        Parameters
        ----------
        w0p : np.double
            Natural radian frequency of the PLL [rad/s]
        w0f : np.double
            Natural radian frequency of the FLL [rad/s]
        w0d : np.double
            Natural radian frequency of the DLL [rad/s]
        doppler : np.double
            Initial doppler estimate [rad/s]
        carrier_phase : np.double, optional
            Initial carrier phase esitmate [rad], by default 0.0
        code_phase : np.double, optional
            Initial code phase estimate [chips], by default 0.0
        nominal_carrier_freq : np.double, optional
            Carrier frequency of the true signal [rad/s], by default 2*pi*1575.42e6
        nominal_code_freq : np.double, optional
            Chipping rate of the true signal [chip/s], by default 1.023e6
        intermediate_freq : np.double
            Intermediate frequency of the recorded signal [rad/s], by default 0.0
        T : np.double, optional
            Integration time [s], by default 0.001
        """
        self.w0p = w0p
        self.w0f = w0f
        self.w0d = w0d
        self.T = T
        self.cn0 = cn0
        self.UpdateCarrierAidingCoeff(nominal_code_freq, nominal_carrier_freq)
        self.__reset_A()
        self.__reset_B()
        self.__reset_Q()
        self.__reset_C()
        self.__reset_R()
        self.x = np.asarray([carrier_phase, doppler, 0.0, code_phase, 0.0])
        self.u = np.asarray([intermediate_freq, nominal_code_freq])
        self.P = np.diag([np.pi**2/3, 2*np.pi*500.0, 100.0, 0.1, 10.0])
        return
            
    def Run(self, phase_err: np.double, freq_err: np.double, chip_err: np.double):
        """
        Run the Kalman Filter DLL/PLL

        Parameters
        ----------
        phase_err : np.double
            Phase discriminator [rad]
        freq_err : np.double
            Frequency discriminator [rad/s]
        chip_err : np.double
            Chip discriminator [chips]
        
        Returns
        -------
        x : np.ndarray
            Current state estimates of PLL/DLL Kalman Filter
        """
        # Predicition
        self.x = self.A @ self.x + self.B @ self.u
        self.P = self.A @ self.P @ self.A.T + self.Q
        
        # Correction
        dy = np.asarray([phase_err, freq_err, chip_err])
        PCt = self.P @ self.C.T
        K = PCt @ np.linalg.inv(self.C @ PCt + self.R)
        self.P = (np.eye(5) - K @ self.C) @ self.P
        self.x += K @ dy
        
        return self.x
       
    def UpdateDoppler(self, doppler: np.double):
        """
        Update tracking loop carrier-to-noise density ratio

        Parameters
        ----------
        cn0 : np.double
            Carrier-to-noise density ratio magnitude (not dB-Hz)
        """
        self.x[1] = doppler
        return
         
    def UpdateCn0(self, cn0: np.double):
        """
        Update tracking loop carrier-to-noise density ratio

        Parameters
        ----------
        cn0 : np.double
            Carrier-to-noise density ratio magnitude (not dB-Hz)
        """
        self.cn0 = cn0
        self.__reset_R()
        return
    
    def UpdateIntegrationTime(self, T: np.double):
        """
        Update tracking loop integration time

        Parameters
        ----------
        T : np.double, optional
            Integration time [s]
        """
        self.T = T
        self.__reset_A()
        self.__reset_B()
        self.__reset_C()
        self.__reset_Q()
        # self.__reset_R()
        # self.P = np.diag([np.pi**2/3, 2*np.pi*500.0, 100.0, 0.1, 10.0])
        return
    
    def UpdateCarrierAidingCoeff(self, 
                                 nominal_code_freq: np.double, 
                                 nominal_carrier_freq: np.double):
        """
        Update coefficient for carrier aiding inside the DLL filter

        Parameters
        ----------
        nominal_code_freq : np.double
            Code frequency (e.g. 1.023e6 for GPS L1 C/A) [Hz]
        nominal_carrier_freq : np.double
            Carrier frequency (e.g. 1575.42e6 for GPS L1 C/A) [Hz]
        """
        self.kappa = nominal_code_freq / nominal_carrier_freq
        return
    
    def UpdateNaturalFreqs(self, w0p: np.double, w0f: np.double, w0d: np.double):
        """
        Update the natural frequencies for the equivalent loop filters

        Parameters
        ----------
        w0p : np.double
            Natural radian frequency of the PLL [rad/s]
        w0f : np.double
            Natural radian frequency of the FLL [rad/s]
        w0d : np.double
            Natural radian frequency of the DLL [rad/s]
        """
        
        self.w0p = w0p
        self.w0f = w0f
        self.w0d = w0d
        self.__reset_Q()
        return
            
    def __reset_A(self):
        T, k = self.T, self.kappa
        self.A = np.asarray(
            [
                [1.0,   T,   0.5*T**2, 0.0, 0.0],
                [0.0, 1.0,          T, 0.0, 0.0],
                [0.0, 0.0,        1.0, 0.0, 0.0],
                [0.0, k*T, 0.5*k*T**2, 1.0,   T],
                [0.0, 0.0,        0.0, 0.0, 1.0]
            ], 
            dtype=np.double
        )
        return
    
    def __reset_B(self):
        self.B = np.asarray(
            [
                [self.T,    0.0],
                [   0.0,    0.0],
                [   0.0,    0.0],
                [   0.0, self.T],
                [   0.0,    0.0]
            ],
            dtype=np.double
        )
        return
    
    def __reset_C(self):
        T, k = self.T, self.kappa
        self.C = np.asarray(
            [
                [1.0,   T/2,   T**2/6, 0.0, 0.0], 
                [0.0,   1.0,      T/2, 0.0, 0.0], 
                [0.0, k*T/2, k*T**2/6, 1.0, T/2]
            ],
            dtype=np.double
        )
        # self.C = np.asarray(
        #     [
        #         [1.0, 0.0, 0.0, 0.0, 0.0], 
        #         [0.0, 1.0, 0.0, 0.0, 0.0], 
        #         [0.0, 0.0, 0.0, 1.0, 0.0]
        #     ],
        #     dtype=np.double
        # )
        return
        
    def __reset_Q(self):
        T, w0d, w0p, w0f, k = self.T, self.w0d, self.w0p, self.w0f, self.kappa
        self.Q = np.asarray(
            [
                [     (w0f*T**5)/20 + (w0p*T**3)/3,   (T**2*(w0f*T**2 + 4*w0p))/8,   (T**3*w0f)/6,                  (T**3*k*(3*w0f*T**2 + 20*w0p))/60,            0],
                [      (w0f*T**4)/8 + (w0p*T**2)/2,          (w0f*T**3)/3 + w0p*T,   (T**2*w0f)/2,                      (T**2*k*(w0f*T**2 + 4*w0p))/8,            0],
                [                     (T**3*w0f)/6,                  (T**2*w0f)/2,          T*w0f,                                     (T**3*k*w0f)/6,            0],
                [(T**3*k*(3*w0f*T**2 + 20*w0p))/60, (T**2*k*(w0f*T**2 + 4*w0p))/8, (T**3*k*w0f)/6, (T**3*(3*w0f*T**2*k**2 + 20*w0p*k**2 + 20*w0d))/60, (T**2*w0d)/2],
                [                                0,                             0,              0,                                       (T**2*w0d)/2,        T*w0d]
            ],
            dtype=np.double
        )
        return
    
    def __reset_R(self):
        self.R = np.asarray(
            [
                [PllVariance(self.cn0, self.T), 0.0, 0.0], # [rad^2]
                [0.0, FllVariance(self.cn0, self.T), 0.0], # [(rad/s)^2]
                [0.0, 0.0, DllVariance(self.cn0, self.T)]  # [chip^2]
            ],
            dtype=np.double
        )
        return

File: src/tracking.cpp
/**
 * *tracking.cpp*
 *
 * =======  ========================================================================================
 * @file    sturdr/tracking.cpp
 * @brief   Satellite scalar tracking methods.
 * @date    January 2025
 * @author  Daniel Sturdivant <sturdivant20@gmail.com>
 * @ref     1. "Understanding GPS/GNSS Principles and Applications", 3rd Edition, 2017
 *              - Kaplan & Hegarty
 *          2. "Are PLLs Dead? A Tutorial on Kalman Filter-Based Techniques for Digital Carrier
 *              Syncronization" - Vila-Valls, Closas, Navarro, Fernandez-Prades
 *          3. "A Software-Defined GPS and Galileo Receiver: A Single-Frequency Approach", 2007
 *              - Borre, Akos, Bertelsen, Rinder, Jensen
 *          4. "Position, Navigation, and Timing Technologies in the 21st Century", Volume 1, 2021
 *              - Morton, Diggelen, Spilker Jr., Parkinson
 * =======  ========================================================================================
 */

#include "sturdr/tracking.hpp"

#include <spdlog/spdlog.h>

#include <cmath>
#include <exception>
#include <navtools/constants.hpp>

#include "sturdr/discriminator.hpp"

namespace sturdr {

inline static const double a = 1.4142;
inline static const double b = 1.1;
inline static const double c = 2.4;
inline static const Eigen::Matrix<double, 5, 5> I5 = Eigen::Matrix<double, 5, 5>::Identity();

// *=== NaturalFrequency ===*
double NaturalFrequency(const double &bw, const int order) {
  try {
    if (order == 1) {
      return bw / 4.0;
    } else if (order == 2) {
      return bw * 8.0 / 3.0;
    } else if (order == 3) {
      return bw * 6.56 / 5.146;
    } else {
      throw std::invalid_argument("Input order must be 1,2, or 3!");
    }

  } catch (std::exception &e) {
    spdlog::get("sturdr-console")
        ->error("tracking.cpp NaturalFrequency failed! Error -> {}", e.what());
    return std::nan("1");
  }
}

// *=== FLLAssistedPLL_2ndOrder ===*
void FLLassistedPLL_2ndOrder(
    double &nco_freq,
    double &vel_accum,
    const double &phase_err,
    const double &freq_err,
    const double &T,
    const double &w0p,
    const double &w0f) {
  try {
    double vel_accum_k = vel_accum + T * (phase_err * w0p * w0p + freq_err * w0f);
    nco_freq = 0.5 * (vel_accum_k + vel_accum) + (phase_err * a * w0p);
    vel_accum = vel_accum_k;

  } catch (std::exception &e) {
    spdlog::get("sturdr-console")
        ->error("tracking.cpp FLLassistedPLL_2ndOrder failed! Error -> {}", e.what());
  }
}

// *=== FLLAssistedFll_3rdOrder ===*
void FLLassistedPLL_3rdOrder(
    double &nco_freq,
    double &vel_accum,
    double &acc_accum,
    const double &phase_err,
    const double &freq_err,
    const double &T,
    const double &w0p,
    const double &w0f) {
  try {
    double w0p2 = w0p * w0p;
    double acc_accum_k = acc_accum + T * (w0p2 * w0p * phase_err + w0f * w0f * freq_err);
    double vel_accum_k = vel_accum + T * (0.5 * (acc_accum_k + acc_accum) + b * w0p2 * phase_err +
                                          a * w0f * freq_err);
    nco_freq = 0.5 * (vel_accum_k + vel_accum) + c * w0p * phase_err;
    vel_accum = vel_accum_k;
    acc_accum = acc_accum_k;

  } catch (std::exception &e) {
    spdlog::get("sturdr-console")
        ->error("tracking.cpp FLLassistedPLL_3rdOrder failed! Error -> {}", e.what());
  }
}

// *=== PLL_2ndOrder ===*
void PLL_2ndOrder(
    double &nco_freq,
    double &vel_accum,
    const double &phase_err,
    const double &T,
    const double &w0) {
  try {
    double vel_accum_k = vel_accum + T * w0 * w0 * phase_err;
    nco_freq = 0.5 * (vel_accum_k + vel_accum) + a * w0 * phase_err;
    vel_accum = vel_accum_k;

  } catch (std::exception &e) {
    spdlog::get("sturdr-console")->error("tracking.cpp PLL_2ndOrder failed! Error -> {}", e.what());
  }
}

// *=== PLL_3rdOrder ===*
void PLL_3rdOrder(
    double &nco_freq,
    double &vel_accum,
    double &acc_accum,
    const double &phase_err,
    const double &T,
    const double &w0) {
  try {
    double w02 = w0 * w0;
    double acc_accum_k = acc_accum + T * w02 * w0 * phase_err;
    double vel_accum_k = vel_accum + T * (0.5 * (acc_accum_k + acc_accum) + b * w02 * phase_err);
    nco_freq = 0.5 * (vel_accum_k + vel_accum) + c * w0 * phase_err;
    vel_accum = vel_accum_k;
    acc_accum = acc_accum_k;

  } catch (std::exception &e) {
    spdlog::get("sturdr-console")->error("tracking.cpp PLL_3rdOrder failed! Error -> {}", e.what());
  }
}

//! ------------------------------------------------------------------------------------------------

// *=== TrackingKF ===*
TrackingKF::TrackingKF()
    : P_{Eigen::Matrix<double, 5, 5>::Zero()},
      Q_{Eigen::Matrix<double, 5, 5>::Zero()},
      R_{Eigen::Matrix<double, 3, 3>::Zero()},
      F_{Eigen::Matrix<double, 5, 5>::Identity()},
      G_{Eigen::Matrix<double, 5, 2>::Zero()},
      H_{Eigen::Matrix<double, 3, 5>::Zero()} {};

// *=== ~TrackingKF ===*
TrackingKF::~TrackingKF(){};

void TrackingKF::Init(
    const double &init_carr_phase,
    const double &init_carr_doppler,
    const double &init_code_phase,
    const double &intmd_freq,
    const double &code_freq) {
  // Initialize state and covariance
  x_ << init_carr_phase, init_carr_doppler, 0.0, init_code_phase, 0.0;
  // P_.diagonal() << navtools::PI_SQU<double> / 3.0, navtools::PI_SQU<> * 125e3, 1000.0, 1.0,
  // 100.0;
  P_.diagonal() << navtools::PI_SQU<double> / 3.0, navtools::PI_SQU<> * 62.5e3,
      navtools::PI_SQU<> * 1e3, 5e-1, 5e2;

  // Initialize dynamic model
  // make_F(init_k, init_T);
  // make_G(init_T);
  // make_Q(init_w0d, init_w0p, init_w0f, init_k, init_T);
  u_ << intmd_freq, code_freq;

  // Initialize measurement model
  make_H();
  // make_R(init_cno, init_T);
}

// *=== UpdateDynamicsParam ===*
void TrackingKF::UpdateDynamicsParam(
    const double &w0d, const double &w0p, const double &w0f, const double &k, const double &T) {
  make_F(k, T);
  make_G(T);
  make_Q(w0d, w0p, w0f, k, T);
}

// *=== UpdateMeasurementsParam ===*
void TrackingKF::UpdateMeasurementsParam(
    const double &dll_var, const double &pll_var, const double &fll_var) {
  // make_H();
  // make_R(cno, T);
  R_.diagonal() << pll_var, fll_var, dll_var;
}

// *=== Run ===*
void TrackingKF::Run(double &chip_err, double &phase_err, double &freq_err) {
  try {
    // Prediction
    x_ = F_ * x_ + G_ * u_;
    P_ = F_ * P_ * F_.transpose() + Q_;

    // Correction
    Eigen::Vector<double, 3> dz{phase_err, freq_err, chip_err};
    Eigen::Matrix<double, 5, 3> PHt = P_ * H_.transpose();
    Eigen::Matrix<double, 5, 3> K = PHt * (H_ * PHt + R_).inverse();
    P_ = (I5 - K * H_) * P_;
    x_ += K * dz;
  } catch (std::exception &e) {
    spdlog::get("sturdr-console")
        ->error("tracking.cpp TrackingKF::Run failed! Error -> {}", e.what());
    // return std::nan("1") * Eigen::Vector<double, 5>::Ones();
  }

  // return x_;
}

void TrackingKF::SetRemCarrierPhase(const double &rem_carrier_phase) {
  x_(0) = rem_carrier_phase;
}
void TrackingKF::SetRemCodePhase(const double &rem_code_phase) {
  x_(3) = rem_code_phase;
}
double TrackingKF::GetRemCarrierPhase() {
  return x_(0);
}
double TrackingKF::GetRemCodePhase() {
  return x_(3);
}
double TrackingKF::GetDoppler() {
  return x_(1);
}

// *=== make_F ===*
void TrackingKF::make_F(const double &k, const double &T) {
  // double T2 = T * T;
  // clang-format off
  // F_ << 1.0,     T,     0.5 * T2, 0.0, 0.0,
  //       0.0,   1.0,            T, 0.0, 0.0,
  //       0.0,   0.0,          1.0, 0.0, 0.0,
  //       0.0, k * T, 0.5 * k * T2, 1.0,   T,
  //       0.0,   0.0,          0.0, 0.0, 1.0;
  // clang-format on
  F_(0, 1) = T;
  F_(0, 2) = 0.5 * T * T;
  F_(1, 2) = T;
  F_(3, 1) = k * T;
  F_(3, 2) = 0.5 * F_(3, 1) * T;
  F_(3, 4) = T;
}

// *=== make_F ===*
void TrackingKF::make_G(const double &T) {
  // clang-format off
  // G_ <<   T, 0.0,
  //       0.0, 0.0,
  //       0.0, 0.0,
  //       0.0,   T,
  //       0.0, 0.0;
  // clang-format on
  G_(0, 0) = T;
  G_(3, 1) = T;
}

// *=== make_Q ===*
void TrackingKF::make_Q(
    const double &w0d, const double &w0p, const double &w0f, const double &k, const double &T) {
  double T2 = T * T;
  double T3 = T * T2;
  // double T4 = T * T3;
  // double T5 = T * T4;
  // double k2 = k * k;
  // clang-format off
  // Q_ <<       (w0f*T5)/20. + (w0p*T3)/3.,   (T2*(w0f*T2 + 4.*w0p))/8.,   (T3*w0f)/6.,               (T3*k*(3.*w0f*T2 + 2.0*w0p))/60.,          0.,
  //              (w0f*T4)/8. + (w0p*T2)/2.,         (w0f*T3)/3. + w0p*T,   (T2*w0f)/2.,                    (T2*k*(w0f*T2 + 4.*w0p))/8.,          0.,
  //                            (T3*w0f)/6.,                 (T2*w0f)/2.,         T*w0f,                                  (T3*k*w0f)/6.,          0.,
  //       (T3*k*(3.*w0f*T2 + 20.*w0p))/60., (T2*k*(w0f*T2 + 4.*w0p))/8., (T3*k*w0f)/6., (T3*(3.*w0f*T2*k2 + 20.*w0p*k2 + 20.*w0d))/60., (T2*w0d)/2.,
  //                                     0.,                          0.,            0.,                                    (T2*w0d)/2.,       T*w0d;
  // clang-format on
  Q_(0, 0) = T3 * ((w0f * T2) / 20.0 + w0p / 3.0);
  Q_(0, 1) = T2 * ((w0f * T2) / 8.0 + w0p / 2.0);
  Q_(0, 2) = (T3 * w0f) / 6.0;
  Q_(0, 3) = (T3 * k * (3.0 * w0f * T2 + 2.0 * w0p)) / 60.0;
  Q_(1, 0) = Q_(0, 1);
  Q_(1, 1) = (w0f * T3) / 3. + w0p * T;
  Q_(1, 2) = (T2 * w0f) / 2.0;
  Q_(1, 3) = (T2 * k * (w0f * T2 + 4. * w0p)) / 8.0;
  Q_(2, 0) = Q_(0, 1);
  Q_(2, 1) = Q_(1, 2);
  Q_(2, 2) = T * w0f;
  Q_(2, 3) = (T3 * k * w0f) / 6.0;
  Q_(3, 0) = Q_(0, 3);
  Q_(3, 1) = Q_(1, 3);
  Q_(3, 2) = Q_(2, 3);
  Q_(3, 3) = (T3 * k * k * (3.0 * w0f * T2 + 20.0 * (w0p + w0d))) / 60.0;
  Q_(3, 4) = (T2 * w0d) / 2.0;
  Q_(4, 3) = Q_(3, 4);
  Q_(4, 4) = T * w0d;
}

// *=== make_H ===*
void TrackingKF::make_H() {
  // // clang-format off
  // H_ << 1.0, 0.0, 0.0, 0.0, 0.0,
  //       0.0, 1.0, 0.0, 0.0, 0.0,
  //       0.0, 0.0, 0.0, 1.0, 0.0;
  // // clang-format on
  H_(0, 0) = 1.0;
  H_(1, 1) = 1.0;
  H_(2, 3) = 1.0;
}

// *=== make_R ===*
void TrackingKF::make_R(const double &cno, const double &T) {
  R_.diagonal() << PllVariance(cno, T), FllVariance(cno, T), DllVariance(cno, T);
}

}  // end namespace sturdr